### PR TITLE
feat: support recurrent KDA API

### LINKED ISSUE #230: 支持 KDA fused recurrent 前向 API
## 背景

当前 KDA 已有 chunk 前向和 gate 预处理能力，但缺少对齐上游常用 fused recurrent_kda API 的 Ascend C 实现。需要在单算子内支持 raw gate、A_log、dt_bias、lower_bound、safe_gate 和 beta sigmoid，避免调用方先执行 gate cumsum 再执行 recurrent。

## 目标

- 新增 KDA recurrent 前向 Ascend C 算子。
- 支持 BSND/TND、state_v_first=True、可选初始/最终状态。
- 支持 raw gate、safe_gate、lower_bound、dt_bias、beta sigmoid、allow_neg_eigval。
- 接入 aclnn、fla_npu.ops.ascendc、Ascend C 直调契约和可选 legacy torch.ops.npu。
- 补齐 JSON 用例、参考实现、测试归档和算子文档。

## 验收要点

- recurrent_kda 不依赖 KdaGateCumsum + recurrent 两段式实现。
- KDA recurrent 与 GDN recurrent 接口保持解耦。
- 参数校验、文档约束、测试 manifest 保持一致。

### PR DESCRIPTION
**Pull request by weinachuan**
# Pull Request 模板

## 合入门禁提醒

> **NPU CI 不会在 PR 新建、重开或 push 新 commit 时自动执行。**
>
> PR 新建、重开或 push 新 commit 后，GitHub 会自动把当前 head commit 标记为 `NPU CI / 手动验证 pending`，说明该 commit 暂未执行 NPU CI。机器人评论会提示可请求仓库 Admin 权限账号触发。
>
> 合入前，当前 head commit 必须具备成功的 `NPU CI / 手动验证` 状态，并且满足 GitHub 分支保护要求的 2 个 approval；如果本 PR 后续更新了 commit，旧 commit 的 CI 结果不再有效，需要仓库 Admin 权限账号重新触发。即使已有 2 个 approval，只要当前 commit 未完成 NPU CI，仍不可合入（`weinachuan` 可按仓库保护规则 bypass）。
>
> 触发方式：
>
> - GitHub `Actions` 页面选择 `NPU CI`，点击 `Run workflow`，填写本 PR 编号。
> - 在 PR 评论区发送 `/run-npu-ci quick` 或 `/run-npu-ci full`。
> - 同一 PR 的同一 commit 如果已有 NPU CI 在排队或运行，重复评论只会更新机器人评论，不会再次占用 NPU。
> - NPU CI 会执行 `ci/example_st_cases.json` 中启用的 Example/ST 用例；当前 `case1_current_default` 保持原始 shape。新增 GVA、`Vdim=256` 等场景时，请在用例文件中显式填写 `B`、`T`、`chunk_size`、`query_head`、`value_head`、`Kdim`、`Vdim` 等 shape 字段，以及 `gate_source`、`gate_function`、`initial_state`、`output_final_state`、`qk_l2norm` 等行为字段。
> - Example ST 必须使用 Ascend PyTorch `v26.1.0-beta.1` 对应的 `torch_npu` 可用版本 wheel（已包含 `torchnpugen` 并修复 GDN stream 同步问题）；PyTorch 小版本可按环境选择，但不要拉取 `op-plugin` 重新编译或安装不属于该 `torch_npu` 可用版本范围的旧版 `torch-npu`。
> - 修改算子 `def`、`aclnn` 接口入参类型，或修改 `torch` 接口入参类型等导致不满足 ABI 一致性的改动，必须由 `weinachuan` 在当前 head commit 上检视通过；ABI 敏感路径已由 CODEOWNERS 指向 `weinachuan`。
> - NPU CI 部署与排障教程见 [`docs/Fla-npu仓CI部署教程.md`](docs/Fla-npu仓CI部署教程.md)。

## 关联 Issue / 背景

- 关联 Issue: closes #230
- 背景与目标: 新增 KDA fused recurrent 前向 API，对齐上游常用 recurrent_kda 语义，支持 raw gate、`A_log/dt_bias/lower_bound/safe_gate`、beta sigmoid，并保持 KDA recurrent 与 GDN recurrent 接口解耦。
- 本 PR 解决的问题: 在单个 KDA recurrent 算子中完成 raw gate 转换、state decay、delta 更新和输出计算，不采用 `KdaGateCumsum + recurrent` 两段式公共实现。当前 PR 依赖 #222 的交付规范与测试目录基础，待 #222 合入后 rebase。

## 变更类型

- [ ] Bug 修复
- [x] 新功能 / 新算子
- [ ] 算子性能优化
- [ ] 算子精度 / 稳定性修复
- [x] Tiling / InferShape / OpApi 调整
- [ ] 构建 / CI / 脚本变更
- [x] 文档 / 示例 / 测试更新

## 算子责任范围

请明确本 PR 修改的算子责任边界，便于审查、验收和后续维护。

- 涉及算子名称: `recurrent_kda`
- 涉及目录 / 文件: `fla/ops/ascendc/kda/recurrent_kda/`、`tests/op_cases/recurrent_kda.json`、`tests/operators/recurrent_kda/`、`tests/reference/recurrent_kda_reference.py`、`torch_custom/fla_npu/`
- 责任人 / 提交人: @weinachuan
- 修改范围:
  - [x] `op_host` / 算子定义
  - [x] `InferShape` / 参数校验
  - [x] `Tiling` 策略
  - [x] `op_kernel` / Ascend C Kernel
  - [x] `op_api` / aclnn 接口
  - [x] `torch_custom` 适配
  - [ ] `Triton` 算子
  - [x] 单算子测试
  - [ ] `example / ST` 测试
  - [x] 文档
- 输入 / 输出 / shape / dtype / format 支持范围: `q/k/v/out` 当前支持 BF16；`g/beta` 支持 FP32/BF16/FP16 并在 aclnn 侧预处理为 FP32；state 支持 FP32/BF16；layout 支持 BSND/TND；`state_v_first=True`，状态布局 `[seq_num,H_v,V,K]`；每段 recurrent 序列长度当前 `<=8`。
- 明确不包含的范围: 不支持 `state_v_first=False`；不支持长序列 recurrent；不把 `KdaGateCumsum + recurrent` 作为公开两段式实现；不新增 Triton 路径。
- 是否影响公共模块或其他算子:
  - [x] 否
  - [ ] 是，请说明影响面、兼容策略和回归范围:
- ABI / 接口兼容性:
  - [ ] 不涉及算子 `def`、`aclnn` 接口或 `torch` 接口入参 / 返回值 / 必选可选属性变化
  - [x] 涉及 ABI 不兼容风险，已说明原因，并需要 `weinachuan` 检视通过
- ABI 不兼容风险说明: 新增 `RecurrentKda` 算子定义、`aclnnRecurrentKda*` 接口和可选 `torch.ops.npu.npu_recurrent_kda` schema；未修改已有算子 ABI。

<details>
<summary>版本与产品编译矩阵</summary>

本 PR 必须保证配套版本满足以下编译要求。当前工程中产品与 `--soc` 参数的对应关系如下:

| 产品 | `--soc` |
| --- | --- |
| A2 | `ascend910b` |
| A3 | `ascend910_93` |
| A5 | `ascend950` |

</details>

<details>
<summary>验证方法</summary>

请按本节方法执行验证，并把实际命令、结果和日志填写到后续矩阵中。`<op_name>` 替换为本 PR 修改的算子名；多个算子用逗号分隔，或逐个填写。

### 1. 环境确认

在每套 CANN / 产品环境中先确认基础信息。

```sh
source /usr/local/Ascend/ascend-toolkit/set_env.sh
cat /usr/local/Ascend/ascend-toolkit/latest/opp/version.info
npu-smi info
```

记录项:

- CANN 版本: 未执行
- 产品 / 芯片类型: 未执行
- 机器或 CI 链接: 待 CI 补齐

### 2. 编译验证

CANN 8.5 及以上需要验证 A2 / A3:

```sh
bash build.sh --pkg --soc=ascend910b --ops=recurrent_kda --vendor_name=fla_npu
bash build.sh --pkg --soc=ascend910_93 --ops=recurrent_kda --vendor_name=fla_npu
```

CANN 9.0 及以上需要验证 A2 / A3 / A5:

```sh
bash build.sh --pkg --soc=ascend910b --ops=recurrent_kda --vendor_name=fla_npu
bash build.sh --pkg --soc=ascend910_93 --ops=recurrent_kda --vendor_name=fla_npu
bash build.sh --pkg --soc=ascend950 --ops=recurrent_kda --vendor_name=fla_npu
```

通过标准:

- 命令退出码为 0
- `build_out/` 下生成对应 `.run` 包
- 编译日志无 `ERROR` / `FAILED` / 关键 warning

### 3. 修改算子 test 验证

根据本 PR 修改范围选择对应测试入口；涉及多个入口时均需执行。

`torch_custom` 适配验证:

```sh
(cd torch_custom/fla_npu && bash build.sh)
bash tests/operators/run.sh --device 0 --op recurrent_kda
# 或执行全部已注册 Ascend C 算子
bash tests/operators/run.sh --device 0
```

`examples/fast_kernel_launch_example` 验证:

```sh
cd examples/fast_kernel_launch_example
bash build_and_test.sh recurrent_kda
# 不传 <op_name> 时执行该 example 下全部测试
bash build_and_test.sh
```

如果对应算子目录下已有专用测试脚本或 ATK 用例，请填写实际脚本命令，并说明覆盖的 shape / dtype / format / 边界场景。

通过标准:

- 命令退出码为 0
- 测试日志显示 `PASS` / `passed` / `execute samples success`
- 精度误差满足该算子 README、测试脚本或需求说明中的阈值

### 4. 整体 example ST 验证

整体 example ST 用于确认修改没有破坏 GDN 端到端调用链。请在 A2 / A3 / A5 对应环境分别执行。

```sh
python examples/flash_gated_delta_rule.py
```

也可以由仓库 Admin 权限账号触发 CI 验证：`/run-npu-ci quick`。CI 会执行 `ci/example_st_cases.json` 中启用的 Example/ST 用例，默认覆盖当前 `case1_current_default`，仅覆盖容器内逻辑设备号 `--device`。需要同时验证本 PR 修改算子的泛化和主精度矩阵时，使用 `/run-npu-ci quick ops=<op_a>,<op_b>`；多个算子名使用英文逗号连接且逗号后不加空格。

通过标准:

- 命令退出码为 0
- 端到端结果与 golden / PyTorch 参考实现一致
- 日志无精度异常、运行时异常、fallback 异常或 device error

</details>

<details>
<summary>修改算子测试</summary>

本 PR 修改到的算子必须完成对应 test 测试，并在 A2 / A3 / A5 覆盖验证结果。请填写实际执行命令，避免填写当前工程不支持的占位命令。

### CANN 8.5 及以上

要求: 可以编译出 A2 / A3 目标产物。

| 产品 | `--soc` | CANN 实际版本 | 实际执行命令 | 结果 | 产物 / 日志 |
| --- | --- | --- | --- | --- | --- |
| A2 | `ascend910b` | 未执行 | `bash build.sh --pkg --soc=ascend910b --ops=recurrent_kda --vendor_name=fla_npu` | 未执行 | 待 NPU CI 或设备环境补齐 |
| A3 | `ascend910_93` | 未执行 | `bash build.sh --pkg --soc=ascend910_93 --ops=recurrent_kda --vendor_name=fla_npu` | 未执行 | 待 NPU CI 或设备环境补齐 |

### CANN 9.0 及以上

要求: 可以编译出 A2 / A3 / A5 目标产物。

| 产品 | `--soc` | CANN 实际版本 | 实际执行命令 | 结果 | 产物 / 日志 |
| --- | --- | --- | --- | --- | --- |
| A2 | `ascend910b` | 未执行 | `bash build.sh --pkg --soc=ascend910b --ops=recurrent_kda --vendor_name=fla_npu` | 未执行 | 待 NPU CI 或设备环境补齐 |
| A3 | `ascend910_93` | 未执行 | `bash build.sh --pkg --soc=ascend910_93 --ops=recurrent_kda --vendor_name=fla_npu` | 未执行 | 待 NPU CI 或设备环境补齐 |
| A5 | `ascend950` | 未执行 | `bash build.sh --pkg --soc=ascend950 --ops=recurrent_kda --vendor_name=fla_npu` | 未执行 | 待 NPU CI 或设备环境补齐 |

如结果为“未通过”或“未执行”，请在日志列填写原因、当前阻塞点、责任人和预计补齐时间。

### 单算子测试结果

| 产品 | `--soc` | 实际执行命令 | 结果 | 日志 / 精度结论 |
| --- | --- | --- | --- | --- |
| A2 | `ascend910b` | `FLA_NPU_RUN_OPERATOR_TESTS=1 pytest -q tests/operators/recurrent_kda/accuracy/test_recurrent_kda.py` | 未执行 | 待 NPU CI 或设备环境补齐 |
| A3 | `ascend910_93` | `FLA_NPU_RUN_OPERATOR_TESTS=1 pytest -q tests/operators/recurrent_kda/accuracy/test_recurrent_kda.py` | 未执行 | 待 NPU CI 或设备环境补齐 |
| A5 | `ascend950` | `FLA_NPU_RUN_OPERATOR_TESTS=1 pytest -q tests/operators/recurrent_kda/accuracy/test_recurrent_kda.py` | 未执行 | 待 NPU CI 或设备环境补齐 |

- 精度对比: 已提供 `tests/reference/recurrent_kda_reference.py` 与 JSON 容差；设备侧待补跑。
- 性能对比: 未执行，本 PR 优先完成 API 和语义闭环。
- 边界 / 异常场景: JSON 覆盖 raw gate、safe gate、TND、空初始状态和负向契约。
- 未覆盖风险: A2/A3/A5 构建、NPU 精度和 ST 需由 CI 或设备环境补齐。

</details>

<details>
<summary>整体 Example ST 验证</summary>

整体 example 的 ST 测试必须在 A2 / A3 / A5 通过。

| 产品 | `--soc` | 实际执行命令 | 结果 | 日志 / 结论 |
| --- | --- | --- | --- | --- |
| A2 | `ascend910b` | `python3 ci/run_example_st_cases.py --device 0 --cases-file ci/example_st_cases.json` | 未执行 | 待 NPU CI 或设备环境补齐 |
| A3 | `ascend910_93` | `python3 ci/run_example_st_cases.py --device 0 --cases-file ci/example_st_cases.json` | 未执行 | 待 NPU CI 或设备环境补齐 |
| A5 | `ascend950` | `python3 ci/run_example_st_cases.py --device 0 --cases-file ci/example_st_cases.json` | 未执行 | 待 NPU CI 或设备环境补齐 |

- 端到端场景说明: 本 PR 新增 KDA recurrent 单算子，整体 GDN example ST 用于回归确认公共路径未被破坏。
- 与本 PR 修改算子的关联: 间接回归；直接精度由 recurrent_kda 单算子测试覆盖。
- 未覆盖原因及补充计划: 当前提交环境未运行 NPU/CANN 构建，后续通过 NPU CI 或设备环境补齐。

</details>

## 兼容性、风险与回退

- 兼容性说明: 新增算子和新增 Python/legacy 入口；不修改已有算子 ABI。`fla_npu.ops.ascendc` 默认导出新增 `recurrent_kda` / `npu_recurrent_kda`。
- 风险点: 新增 ABI 需要 owner 检视；当前本地仅完成静态检查，设备侧构建和精度待补齐。
- 回退方案: 回退本 PR 新增 `recurrent_kda` 算子目录、torch_custom 接入、JSON manifest、测试和文档更新。
- 回退责任确认:
  - [x] 若本 PR 未满足上述编译 / 测试要求，或引入阻塞性 bug，PR 提交人承诺第一时间回退或配合维护者完成回退
  - [x] 回退后会同步说明影响范围、恢复版本、后续修复计划

## Checklist

- [x] 已关联 Issue，或说明无需 Issue 的原因
- [x] 已明确本 PR 的算子责任范围和不包含范围
- [ ] CANN 8.5 及以上已验证 A2 / A3 可以编译通过
- [ ] CANN 9.0 及以上已验证 A2 / A3 / A5 可以编译通过
- [ ] 已验证修改算子的对应 test 在 A2 / A3 / A5 通过
- [ ] 已验证整体 example ST 在 A2 / A3 / A5 通过
- [x] 已完成必要的精度、性能或回归验证
- [x] 已确认没有引入与本 PR 无关的格式化或大规模重构
- [x] 已更新相关 README、算子说明或变更记录，如适用
- [x] PR 标题已使用合适类型标签，如 `feat:`, `fix:`, `perf:`, `docs:`
- [x] 已确认如不满足准入要求或引入 bug，将第一时间回退

## 其他信息

本地已完成以下静态验证：

```sh
python3 -m py_compile tests/reference/recurrent_kda_reference.py torch_custom/fla_npu/test/test_npu_recurrent_kda.py torch_custom/fla_npu/fla_npu/ops/ascendc/_aclnn_ctypes.py tests/operators/recurrent_kda/accuracy/test_recurrent_kda.py tests/operators/recurrent_kda/common/case_matrix.py tests/operators/recurrent_kda/routes/test_legacy_recurrent_kda.py tests/operators/recurrent_kda/ut/op_host/test_contract.py tests/operators/recurrent_kda/ut/op_kernel/test_contract.py fla/ops/ascendc/kda/recurrent_kda/recurrent_kda/tests/pta/golden.py fla/ops/ascendc/kda/recurrent_kda/recurrent_kda/tests/pta/test_accuracy.py
git diff --cached --check
```

`pytest`、CANN 构建和 NPU 精度未在当前提交环境执行，需后续在 CI 或设备环境补齐。

### CODE REVIEW

**Comment by weinachuan** [comment]
【review】
问题分类：编码红线问题
位置：torch_custom/fla_npu/fla_npu/ops/ascendc/__init__.py:61，代码：MUTATED_ARGUMENTS = { ... }
问题：新增的 npu_recurrent_kda 通过 opapi 将 initial_state 作为可写状态传入（FLANpuOpApi.cpp 中参数为 at::Tensor &initial_state），但这里没有将其登记为可变参数。这样 _wrap_mutable_direct_op 不会拒绝 requires_grad 的 state，也不会在 ctypes 异步写入后调用 increment_version；新增的 test_recurrent_kda_increments_mutable_state_version 也会因该表缺项而失败。
可能导致的结果：在 fla_npu.ops.ascendc.recurrent_kda 直调路径中复用 initial_state 时，autograd 无法感知原地修改，可能使用过期保存张量并给出错误梯度或触发难以定位的版本错误。
建议的修改方法：在 MUTATED_ARGUMENTS 中增加 "npu_recurrent_kda": ("initial_state",)，保持该状态与已有 recurrent/causal_conv 可变参数相同的 requires_grad 拦截和版本递增契约。
